Fix latest plugin version not being downloaded consistently

ShyamGadde opened the pull request:
## Summary

Fixes #1690

ShyamGadde committed:
Remove plugin version number from download link

This is done so that it automatically downloads the latest version

westonruter commented on plugins/performance-lab/includes/admin/plugins.php:
```diff
@@ -372,7 +372,9 @@ function perflab_install_and_activate_plugin( string $plugin_slug, array &$proce
 		$upgrader = new Plugin_Upgrader( $skin );
-		$result   = $upgrader->install( $plugin_data['download_link'] );
+		// Remove the version number from the link to download the latest plugin version.
+		$download_link = (string) preg_replace( '#(\/plugin\/[^\/]+)\.\d+\.\d+\.\d+\.zip#', '$1.zip', $plugin_data['download_link'] );
```
It might be safer to do:

```suggestion
		$download_link = (string) preg_replace( '#(\/plugin\/[^\/]+)\.\d.*.zip#', '$1.zip', $plugin_data['download_link'] );
```

The reason being that it's possible for plugin versions to have `-rc` or `-beta` tagged on. 

For example: 
https://wordpress.org/plugins/rest-api/
=>
https://downloads.wordpress.org/plugin/rest-api.2.0-beta15.zip

Granted, it's highly unlikely that we'll ever release one of our plugins with such a Stable Tag, and it's not super common across the repo: https://wpdirectory.net/search/01JDAFQMKVDE6ME17SH47V33D9

westonruter replied:
So I don't feel strongly about this.

westonruter replied:
Worst case in such a scenario is that the original URL ends up getting used without the version removed, which will likely still work (even if it has the even slighter possibility of downloading a stale version).

ShyamGadde replied:
Thank you for pointing this out! I hadn’t considered the possibility of version tags like `-rc` or `-beta`. That’s an oversight on my part.

However, I believe the regex `'#(\/plugin\/[^\/]+)\.\d.*.zip#'` might not work as intended because it could capture the plugin name greedily. For instance, in the example provided, the download link might end up being `https://downloads.wordpress.org/plugin/rest-api.2.zip` instead of `https://downloads.wordpress.org/plugin/rest-api.zip`.

To address this, one solution could be to simply modify the regex to capture the plugin name non-greedily:

```php
'#(\/plugin\/[^\/]+?)\.\d.*.zip#'
```

Alternatively, another approach would be to use the following regex:

```php
'#(\/plugin\/[^\/]+)\.\d+\.\d+\.\d+(?:-\w+)?\.zip#'
```

This pattern accounts for optional tags like `-rc` or `-beta` and aligns with the regex used in the `versions` command, so it seems like a good choice to me.

westonruter replied:
Note in the example https://downloads.wordpress.org/plugin/rest-api.2.0-beta15.zip there is no patch version either, just a major and minor. So that would need to be optional as well.

This should address the greediness issue, by using `\.\d.*?.zip` instead of `\.\d.*.zip`:

```php
$download_link = (string) preg_replace( '#(\/plugin\/[^\/]+)\.\d.*?.zip#', '$1.zip', $plugin_data['download_link'] );
```

ShyamGadde committed:
Update regex to handle version tags like `-rc` and `-beta` in plugin download links

westonruter commented on plugins/performance-lab/includes/admin/plugins.php:
```diff
@@ -372,7 +372,9 @@ function perflab_install_and_activate_plugin( string $plugin_slug, array &$proce
 		$upgrader = new Plugin_Upgrader( $skin );
-		$result   = $upgrader->install( $plugin_data['download_link'] );
+		// Remove the version number from the link to download the latest plugin version.
+		$download_link = (string) preg_replace( '#(\/plugin\/[^\/]+)\.\d+\.\d+\.\d+(?:-\w+)?\.zip#', '$1.zip', $plugin_data['download_link'] );
```
```suggestion
		$download_link = (string) preg_replace( '#(\/plugin\/[^\/]+)\.\d.*?.zip#', '$1.zip', $plugin_data['download_link'] );
```

westonruter replied:
Or rather, per https://github.com/WordPress/performance/pull/1693#issuecomment-2495000449

```suggestion
		$download_link = (string) preg_replace( '#(\/plugin\/[^\/.]+)\..+?\.zip#', '$1.zip', $plugin_data['download_link'] );
```

ShyamGadde replied:
Stripping off everything up to the first `.` makes sense. However, I also noticed that some plugins have their download link as `https://downloads.wordpress.org/plugin/<plugin-slug>.zip`, particularly those with tags like `trunk`, such as https://downloads.wordpress.org/plugin/better-search-replace.zip.

Given this, I think we might need a regex like this:

```php
$download_link = (string) preg_replace( '#(\/plugin\/[^\.]+).*\.zip#', '$1.zip', $plugin_data['download_link'] );
```

This should account for both cases where the version exists (like `rest-api.2.0-beta15.zip`) and where it doesn't (like `better-search-replace.zip`).

ShyamGadde replied:
On second thought, can’t we just construct the download link directly from the slug in the `$plugin_data` as:

```
https://downloads.wordpress.org/plugin/<plugin-slug>.zip
```

This would simplify and remove the need for regex.

westonruter replied:
Hah. Good point.

westonruter commented:
Lastly, note that many plugins also use `Stable tag: trunk`: https://wpdirectory.net/search/01JDB0D3QBVXR8KWR1X9BRYYPB

I haven't seen `trunk` show up in the download URL, but in theory it might?

Here's another search for plugins using non-standard versions: https://wpdirectory.net/search/01JDB0R6MBNNEMH0A1X8AHKQYC

The WP Umbrella plugin version doesn't begin with a digit:

https://downloads.wordpress.org/plugin/wp-health.v2.16.4.zip

That being said, perhaps we should just strip off everything up to the first `.` since I don't believe any plugin is allowed to have that as a character in the slug.

westonruter commented:
Again, this is all likely splitting hairs since none of the plugins we include here should ever use a non-standard version!

felixarntz commented on plugins/performance-lab/includes/admin/plugins.php:
```diff
@@ -372,7 +372,9 @@ function perflab_install_and_activate_plugin( string $plugin_slug, array &$proce
 		$upgrader = new Plugin_Upgrader( $skin );
-		$result   = $upgrader->install( $plugin_data['download_link'] );
+		// Remove the version number from the link to download the latest plugin version.
+		$download_link = (string) preg_replace( '#(\/plugin\/[^\/]+)\.\d+\.\d+\.\d+(?:-\w+)?\.zip#', '$1.zip', $plugin_data['download_link'] );
```
I'm not a fan of regexing this from the `download_link`. Is there no externally controlled place (API responses or elsewhere) that exposes the format we want here?

felixarntz replied:
Seeing comments like https://github.com/WordPress/performance/pull/1693#discussion_r1856636899 confirms my above concern: Using a regex for this is error-prone. Is there no way to get the actual URL from somewhere?

ShyamGadde replied:
You’re right; relying on regex for this isn’t ideal. The `plugins_api()` function, when called with the `plugin_information` action, provides a `versions` field in the response that seems to include the URLs in the required format. Here's an example of the response:

```json
{
    ...
    "versions": {
        "0.1.0": "https://downloads.wordpress.org/plugin/optimization-detective.0.1.0.zip",
        "0.1.1": "https://downloads.wordpress.org/plugin/optimization-detective.0.1.1.zip",
        ...
        "trunk": "https://downloads.wordpress.org/plugin/optimization-detective.zip"
    },
    ...
}
```

We could use the `trunk` key to get the required URL in most cases. However, there are scenarios where the `trunk` key doesn’t exist in the `versions` field. In those cases, the `download_link` itself is typically already in the desired format without a version or tag.

That said, implementing this approach would require making multiple API calls—one for each plugin—which again might not be ideal.

felixarntz replied:
Don't we only need the plugin download URL if we actually want to download the plugin? I think making an extra request for that purpose makes sense. We shouldn't make all these requests for all the plugins when we cache the information. But for the moment when someone clicks "Activate" and we have to install that plugin, I think it's reasonable to make the request. I think that's even what WordPress Core does to install a plugin.

And even if the plugin also requires dependencies, making individual requests for these should not be an issue since realistically that shouldn't be more than 2-3 plugins.

This solution would be more reliable than us assuming things about the download URL and using a regex.

westonruter replied:
That makes sense to me.

ShyamGadde committed:
Avoid transient cache for plugin installation to ensure latest version

Replaced transient-cached plugin data with a real-time `plugins_api()`
call during installation to avoid installing outdated versions.

westonruter committed:
Remove download_link from data explicitly requested in perflab_query_plugin_info()

westonruter committed:
Explicitly request download_link and requires_plugins fields

westonruter reviewed (approved):
Since we're not using the `download_link` anymore in the response to `perflab_query_plugin_info()`, I removed it from being explicitly returned. I also explicitly requested `download_link` and `requires_plugins` when installing, although it seems these are all returned by default anyway.

westonruter committed:
Explain why perflab_query_plugin_info() isn't used in perflab_install_and_activate_plugin()

felixarntz reviewed (approved):
Thanks @ShyamGadde @westonruter, LGTM!
